README.md
# @dev-ptera/nano-rpc-proxy

This is an express app which filters incoming Nano RPC requests and queries a configured Nano node.  It is intended that this app runs on the same machine as a fully synced Nano node.

For instructions on how to setup a Nano node, visit https://docs.nano.org/running-a-node/node-setup/

## Usage

To install the package:

`yarn add @dev-ptera/nano-rpc-proxy`

or

`npm i @dev-ptera/nano-rpc-proxy`

###
Add the following to your typescript application:
```ts
import * as express from 'express';
import { NanoProxyServer } from '@dev-ptera/nano-rpc-proxy';


const server = new NanoProxyServer(express(), {
    
    /* Server message emitted when app starts listening on `port`. */
    APP_LISTENING_MSG: (port: number) => 
        `Running @dev-ptera/nano-rpc-proxy server on port ${port}.`,

    /* Server is expected to serve external requests. */
    IS_PRODUCTION: false,

    /* Nano/Banano Node RPC URL */
    NANO_RPC_URL: 'http://[::1]:7072',

    /* Server port when ran locally for development purposes. */
    APP_DEV_PORT: 1119,

    /* Server port when listening to outside requests. */
    APP_PROD_PORT: 1120,

    /* URL path where the app is served.  
       Example: http://[YOUR-IP]:[APP_DEV_PORT | APP_PROD_PORT]/[APP_PATH] 
    */
    APP_PATH: '',

    /* 
       List of enabled websites that can bypass server CORS restriction.
       CORS is not enforced when server is ran in development-mode.
    */
    URL_WHITE_LIST: ['http://localhost'],

    /*
       List of actions we can use with Nano RPC Protocol,
       Full list of actions & descriptions here: https://docs.nano.org/commands/rpc-protocol
     */
    ALLOWED_ACTIONS: ['account_balance']

    /*
     * Optional function ran to security-check incoming requests.
     * Runs after the CORS filter.
     */
    REQUEST_FILTER: (req: Request, res: Response, next: NextFunction) => void;
};
});

server.start();
```


## Configuration

This server can be configured to support:

- Whitelisted external domains.
- Enabled/disabled RPC actions
- Nano or Banano Nodes
- Custom Server Ports & Server Path
- etc.

All supported config options can be found in `src/config.ts`


## Error Handling

This sever returns an error code 501 when a disabled action was requested by a client.  
This server returns an error code 500 if a request is blocked by CORS or any other generic error happens within the proxy.


If the Nano RPC request returns an `error` response due to invalid or missing account info, JSON parsing issues, etc this proxy will return a status code 200 since the error was generated from the Nano Node.


CHANGELOG.md
# Changelog

## v1.2.0

### Added

- Added wildcard option to `URL_WHITE_LIST` config option to passby CORS restriction in production mode.

## v1.1.2

### Added

- Added `REQUEST_FILTER` config option for additional security check to incoming requests.

## v1.0.0

### Added
- Added new config prop `APP_LISTENING_MSG`.
- Added all Nano RPC actions. 
- Added comments marking deprecated APIs.

## v0.0.1
Initial Release

src/config.ts
import { NextFunction, Request, Response } from 'express';

export type NanoProxyServerConfig = {
    /** Server message emitted when app starts listening on `port`. */
    APP_LISTENING_MSG: (port: number) => string;

    /** Server is expected to serve external requests. */
    IS_PRODUCTION: boolean;

    /** Nano/Banano Node RPC URL */
    NANO_RPC_URL: string;

    /** Server port when ran locally for development purposes. */
    APP_DEV_PORT: number;

    /** Server port when listening to outside requests. */
    APP_PROD_PORT: number;

    /** URL path where the app is served.  Example: http://[YOUR-IP]:[APP_DEV_PORT | APP_PROD_PORT]/[APP_PATH] */
    APP_PATH: string;

    /**
       List of enabled websites that can bypass server CORS restriction.
       CORS is not enforced when server is ran in development-mode.
    */
    URL_WHITE_LIST: string[];

    /**
       List of actions we can use with Nano RPC Protocol,
       Full list of actions & descriptions here: https://docs.nano.org/commands/rpc-protocol
     */
    ALLOWED_ACTIONS: AllowedActions[];

    /**
     * Optional function ran to security-check incoming requests.
     * Runs after the CORS filter.
     */
    REQUEST_FILTER?: (req: Request, res: Response, next: NextFunction) => void;
};

export type AllowedActions =
    | 'account_balance'
    | 'account_block_count'
    | 'account_count'
    | 'account_create'
    | 'account_get'
    | 'account_history'
    | 'account_info'
    | 'account_key'
    | 'account_list'
    | 'account_move'
    | 'account_remove'
    | 'account_representative'
    | 'account_representative_set'
    | 'account_weight'
    | 'accounts_balances'
    | 'accounts_create'
    | 'accounts_frontiers'
    | 'accounts_pending'
    | 'available_supply'
    | 'block'
    | 'block_account'
    | 'block_confirm'
    | 'block_count'
    | 'block_count_type' /* Removed in V22 */
    | 'block_create'
    | 'block_hash'
    | 'blocks'
    | 'blocks_info'
    | 'successors'
    | 'bootstrap'
    | 'bootstrap_any'
    | 'bootstrap_lazy'
    | 'bootstrap_status'
    | 'chain'
    | 'confirmation_active'
    | 'confirmation_height_currently_processing'
    | 'confirmation_history'
    | 'confirmation_info'
    | 'confirmation_quorum'
    | 'database_txn_tracker'
    | 'delegators'
    | 'delegators_count'
    | 'deterministic_key'
    | 'epoch_upgrade'
    | 'frontier_count'
    | 'frontiers'
    | 'history'
    | 'keepalive'
    | 'key_create'
    | 'key_expand'
    | 'krai_from_raw'
    | 'krai_to_raw'
    | 'ledger'
    | 'node_id'
    | 'node_id_delete'
    | 'mrai_from_raw'
    | 'mrai_to_raw'
    | 'password_change'
    | 'password_enter'
    | 'password_valid'
    | 'payment_begin' /* Removed in V22 */
    | 'payment_init' /* Removed in V22 */
    | 'payment_end' /* Removed in V22 */
    | 'payment_wait' /* Removed in V22 */
    | 'peers'
    | 'pending'
    | 'pending_exists'
    | 'process'
    | 'rai_from_raw'
    | 'rai_to_raw'
    | 'receive'
    | 'receive_minimum'
    | 'receive_minimum_set'
    | 'representatives'
    | 'representatives_online'
    | 'republish'
    | 'search_pending'
    | 'search_pending_all'
    | 'send'
    | 'sign'
    | 'stats'
    | 'stats_clear'
    | 'stop'
    | 'successors'
    | 'telemetry'
    | 'validate_account_number'
    | 'version'
    | 'unchecked'
    | 'unchecked_clear'
    | 'unchecked_get'
    | 'unchecked_keys'
    | 'unopened'
    | 'uptime'
    | 'wallet_add'
    | 'wallet_add_watch'
    | 'wallet_balances'
    | 'wallet_change_seed'
    | 'wallet_contains'
    | 'wallet_create'
    | 'wallet_destroy'
    | 'wallet_export'
    | 'wallet_frontiers'
    | 'wallet_history'
    | 'wallet_info'
    | 'wallet_key_valid'
    | 'wallet_ledger'
    | 'wallet_lock'
    | 'wallet_locked'
    | 'wallet_pending'
    | 'wallet_representative'
    | 'wallet_representative_set'
    | 'wallet_republish'
    | 'wallet_work_get'
    | 'work_cancel'
    | 'work_generate'
    | 'work_get'
    | 'work_peer_add'
    | 'work_peers'
    | 'work_peers_clear'
    | 'work_set'
    | 'work_validate';


src/app.ts
import * as cors from 'cors';
import * as http from 'http';
import * as bodyParser from 'body-parser';
import { NanoProxyServerConfig } from './config';
import axios, { AxiosResponse } from 'axios';
import { Express, Request, Response } from 'express';

export class NanoProxyServer {
    app: Express;
    config: NanoProxyServerConfig;

    constructor(app: Express, config: NanoProxyServerConfig) {
        this.app = app;
        this.config = config;
        const corsOptions = {
            origin: function (origin, callback) {

                if (config.URL_WHITE_LIST[0] == '**') {
                    return callback(null, true);
                }

                if (config.IS_PRODUCTION
                    && origin
                    && config.URL_WHITE_LIST
                    && config.URL_WHITE_LIST.indexOf(origin) === -1) {
                    return callback(new Error(`Origin '${origin}' is not allowed by CORS`));
                } else {
                    return callback(null, true);
                }
            }
        };

        const send = (res, data, status: number): void => res.status(status).send(JSON.stringify(data));
        const contactRpc = (body): Promise<any> =>
            axios
                .request<any>({
                    method: 'POST',
                    url: config.NANO_RPC_URL,
                    data: body,
                })
                .then((response: AxiosResponse) => Promise.resolve(response.data))
                .catch((err) => Promise.reject(err));

        app.use(bodyParser.json());
        app.use(cors(corsOptions));
        if (config.REQUEST_FILTER) {
            app.use(config.REQUEST_FILTER);
        }
        app.post(`/${config.APP_PATH}`, async (req: Request, res: Response) => {
            const body = req.body;
            if (!body || !body.action || !config.ALLOWED_ACTIONS.includes(body.action)) {
                const error = `RPC action not enabled: ${req.body?.action}`;
                return send(res, { error }, 501);
            }
            await contactRpc(body)
                .then((data) => send(res, data, 200))
                .catch((err: Error) => send(res, err, 500));
        });
    }

    start(): void {
        const port = this.config.IS_PRODUCTION ? this.config.APP_PROD_PORT : this.config.APP_DEV_PORT;
        http.createServer(this.app).listen(port, () => {
            console.log(this.config.APP_LISTENING_MSG(port));
        });
    }
}
